File: cn/java-springmvc/src/main/java/com/example/controller/ShippingController.java
package com.example.controller;

import org.springframework.web.bind.annotation.*;
import com.example.model.Shipping;
import com.example.dto.Result;
import com.example.dto.PageResult;
import jakarta.validation.Valid;
import java.math.BigDecimal;
import java.util.List;

@RestController
@RequestMapping("/api/shipping")
public class ShippingController {
    
    @GetMapping
    public Result<PageResult<Shipping>> listShipping(
            @RequestParam(required = false) Long orderId,
            @RequestParam(required = false) String trackingNo,
            @RequestParam(required = false) Integer status,
            @RequestParam(defaultValue = "1") Integer pageNum,
            @RequestParam(defaultValue = "10") Integer pageSize) {
        return Result.success(new PageResult<>());
    }
    
    @PostMapping
    public Result<Shipping> createShipping(@Valid @RequestBody CreateShippingReq req) {
        return Result.success(new Shipping());
    }
    
    @GetMapping("/{id}")
    public Result<Shipping> getShipping(@PathVariable Long id) {
        return Result.success(new Shipping());
    }
    
    @PutMapping("/{id}")
    public Result<Shipping> updateShipping(@PathVariable Long id, @Valid @RequestBody UpdateShippingReq req) {
        return Result.success(new Shipping());
    }
    
    @GetMapping("/no/{shippingNo}")
    public Result<Shipping> getShippingByNo(@PathVariable String shippingNo) {
        return Result.success(new Shipping());
    }
    
    @GetMapping("/{id}/track")
    public Result<ShippingTrackVO> trackShipping(@PathVariable Long id) {
        return Result.success(new ShippingTrackVO());
    }
    
    @PostMapping("/{id}/cancel")
    public Result<Shipping> cancelShipping(@PathVariable Long id) {
        return Result.success(new Shipping());
    }
    
    @GetMapping("/order/{orderId}")
    public Result<Shipping> getShippingByOrder(@PathVariable Long orderId) {
        return Result.success(new Shipping());
    }
    
    @PostMapping("/calculate-cost")
    public Result<ShippingCostVO> calculateCost(@Valid @RequestBody CalculateShippingCostReq req) {
        return Result.success(new ShippingCostVO());
    }
    
    @GetMapping("/carriers")
    public Result<List<CarrierVO>> getCarriers() {
        return Result.success(List.of());
    }
    
    @PutMapping("/{id}/status")
    public Result<Shipping> updateStatus(@PathVariable Long id, @RequestParam Integer status) {
        return Result.success(new Shipping());
    }
    
    @PutMapping("/{id}/tracking")
    public Result<Shipping> updateTracking(
            @PathVariable Long id,
            @RequestParam String trackingNo,
            @RequestParam String carrier) {
        return Result.success(new Shipping());
    }
    
    @GetMapping("/{id}/timeline")
    public Result<List<ShippingTimelineVO>> getTimeline(@PathVariable Long id) {
        return Result.success(List.of());
    }
    
    @GetMapping("/pending")
    public Result<PageResult<Shipping>> getPendingShipping(
            @RequestParam(defaultValue = "1") Integer pageNum,
            @RequestParam(defaultValue = "10") Integer pageSize) {
        return Result.success(new PageResult<>());
    }
    
    @GetMapping("/in-transit")
    public Result<PageResult<Shipping>> getInTransitShipping(
            @RequestParam(defaultValue = "1") Integer pageNum,
            @RequestParam(defaultValue = "10") Integer pageSize) {
        return Result.success(new PageResult<>());
    }
    
    @GetMapping("/delivered")
    public Result<PageResult<Shipping>> getDeliveredShipping(
            @RequestParam(defaultValue = "1") Integer pageNum,
            @RequestParam(defaultValue = "10") Integer pageSize) {
        return Result.success(new PageResult<>());
    }
    
    @GetMapping("/statistics")
    public Result<ShippingStatisticsVO> getStatistics() {
        return Result.success(new ShippingStatisticsVO());
    }
    
    @GetMapping("/export")
    public Result<String> exportShipping(
            @RequestParam(required = false) String startDate,
            @RequestParam(required = false) String endDate) {
        return Result.success("export_file_url");
    }
    
    @PostMapping("/{id}/confirm-delivery")
    public Result<Shipping> confirmDelivery(@PathVariable Long id) {
        return Result.success(new Shipping());
    }
    
    @PostMapping("/{id}/reschedule")
    public Result<Shipping> rescheduleDelivery(
            @PathVariable Long id,
            @RequestParam String newDate) {
        return Result.success(new Shipping());
    }
}

class CreateShippingReq {
    private Long orderId;
    private String carrier;
    private String receiverName;
    private String receiverPhone;
    private String shippingAddress;
    
    public Long getOrderId() { return orderId; }
    public void setOrderId(Long orderId) { this.orderId = orderId; }
    public String getCarrier() { return carrier; }
    public void setCarrier(String carrier) { this.carrier = carrier; }
    public String getReceiverName() { return receiverName; }
    public void setReceiverName(String receiverName) { this.receiverName = receiverName; }
    public String getReceiverPhone() { return receiverPhone; }
    public void setReceiverPhone(String receiverPhone) { this.receiverPhone = receiverPhone; }
    public String getShippingAddress() { return shippingAddress; }
    public void setShippingAddress(String shippingAddress) { this.shippingAddress = shippingAddress; }
}

class UpdateShippingReq {
    private String carrier;
    private String receiverName;
    private String receiverPhone;
    private String shippingAddress;
    
    public String getCarrier() { return carrier; }
    public void setCarrier(String carrier) { this.carrier = carrier; }
    public String getReceiverName() { return receiverName; }
    public void setReceiverName(String receiverName) { this.receiverName = receiverName; }
    public String getReceiverPhone() { return receiverPhone; }
    public void setReceiverPhone(String receiverPhone) { this.receiverPhone = receiverPhone; }
    public String getShippingAddress() { return shippingAddress; }
    public void setShippingAddress(String shippingAddress) { this.shippingAddress = shippingAddress; }
}

class ShippingTrackVO {
    private String shippingNo;
    private String carrier;
    private String trackingNo;
    private Integer status;
    private String currentLocation;
    private List<TrackEventVO> events;
    
    public String getShippingNo() { return shippingNo; }
    public void setShippingNo(String shippingNo) { this.shippingNo = shippingNo; }
    public String getCarrier() { return carrier; }
    public void setCarrier(String carrier) { this.carrier = carrier; }
    public String getTrackingNo() { return trackingNo; }
    public void setTrackingNo(String trackingNo) { this.trackingNo = trackingNo; }
    public Integer getStatus() { return status; }
    public void setStatus(Integer status) { this.status = status; }
    public String getCurrentLocation() { return currentLocation; }
    public void setCurrentLocation(String currentLocation) { this.currentLocation = currentLocation; }
    public List<TrackEventVO> getEvents() { return events; }
    public void setEvents(List<TrackEventVO> events) { this.events = events; }
}

class TrackEventVO {
    private String time;
    private String location;
    private String description;
    
    public String getTime() { return time; }
    public void setTime(String time) { this.time = time; }
    public String getLocation() { return location; }
    public void setLocation(String location) { this.location = location; }
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
}

class CalculateShippingCostReq {
    private String fromAddress;
    private String toAddress;
    private BigDecimal weight;
    private String carrier;
    
    public String getFromAddress() { return fromAddress; }
    public void setFromAddress(String fromAddress) { this.fromAddress = fromAddress; }
    public String getToAddress() { return toAddress; }
    public void setToAddress(String toAddress) { this.toAddress = toAddress; }
    public BigDecimal getWeight() { return weight; }
    public void setWeight(BigDecimal weight) { this.weight = weight; }
    public String getCarrier() { return carrier; }
    public void setCarrier(String carrier) { this.carrier = carrier; }
}

class ShippingCostVO {
    private BigDecimal baseCost;
    private BigDecimal weightCost;
    private BigDecimal totalCost;
    private Integer estimatedDays;
    
    public BigDecimal getBaseCost() { return baseCost; }
    public void setBaseCost(BigDecimal baseCost) { this.baseCost = baseCost; }
    public BigDecimal getWeightCost() { return weightCost; }
    public void setWeightCost(BigDecimal weightCost) { this.weightCost = weightCost; }
    public BigDecimal getTotalCost() { return totalCost; }
    public void setTotalCost(BigDecimal totalCost) { this.totalCost = totalCost; }
    public Integer getEstimatedDays() { return estimatedDays; }
    public void setEstimatedDays(Integer estimatedDays) { this.estimatedDays = estimatedDays; }
}

class CarrierVO {
    private String code;
    private String name;
    private String logo;
    private Boolean enabled;
    
    public String getCode() { return code; }
    public void setCode(String code) { this.code = code; }
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    public String getLogo() { return logo; }
    public void setLogo(String logo) { this.logo = logo; }
    public Boolean getEnabled() { return enabled; }
    public void setEnabled(Boolean enabled) { this.enabled = enabled; }
}

class ShippingTimelineVO {
    private String time;
    private String status;
    private String description;
    private String location;
    
    public String getTime() { return time; }
    public void setTime(String time) { this.time = time; }
    public String getStatus() { return status; }
    public void setStatus(String status) { this.status = status; }
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    public String getLocation() { return location; }
    public void setLocation(String location) { this.location = location; }
}

class ShippingStatisticsVO {
    private Long totalShipping;
    private Long pendingShipping;
    private Long inTransitShipping;
    private Long deliveredShipping;
    private Long cancelledShipping;
    
    public Long getTotalShipping() { return totalShipping; }
    public void setTotalShipping(Long totalShipping) { this.totalShipping = totalShipping; }
    public Long getPendingShipping() { return pendingShipping; }
    public void setPendingShipping(Long pendingShipping) { this.pendingShipping = pendingShipping; }
    public Long getInTransitShipping() { return inTransitShipping; }
    public void setInTransitShipping(Long inTransitShipping) { this.inTransitShipping = inTransitShipping; }
    public Long getDeliveredShipping() { return deliveredShipping; }
    public void setDeliveredShipping(Long deliveredShipping) { this.deliveredShipping = deliveredShipping; }
    public Long getCancelledShipping() { return cancelledShipping; }
    public void setCancelledShipping(Long cancelledShipping) { this.cancelledShipping = cancelledShipping; }
}

File: en/kotlin-springmvc/src/main/kotlin/com/example/controller/ShippingController.kt
package com.example.controller

import com.example.dto.PageResult
import com.example.dto.Result
import com.example.model.Shipping
import org.springframework.web.bind.annotation.*
import java.math.BigDecimal

@RestController
@RequestMapping("/api/shipping")
class ShippingController {
    
    @GetMapping
    fun listShipping(
        @RequestParam(required = false) orderId: Long?,
        @RequestParam(required = false) trackingNo: String?,
        @RequestParam(required = false) status: Int?,
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<Shipping>> {
        val result = PageResult<Shipping>()
        return Result.success(result)
    }
    
    @PostMapping
    fun createShipping(@RequestBody req: CreateShippingReq): Result<Shipping> {
        val shipping = Shipping()
        return Result.success(shipping)
    }
    
    @GetMapping("/{id}")
    fun getShipping(@PathVariable id: Long): Result<Shipping> {
        val shipping = Shipping()
        return Result.success(shipping)
    }
    
    @PutMapping("/{id}")
    fun updateShipping(@PathVariable id: Long, @RequestBody req: UpdateShippingReq): Result<Shipping> {
        val shipping = Shipping()
        return Result.success(shipping)
    }
    
    @DeleteMapping("/{id}")
    fun deleteShipping(@PathVariable id: Long): Result<Void> {
        return Result.success(null)
    }
    
    @GetMapping("/order/{orderId}")
    fun getShippingByOrder(@PathVariable orderId: Long): Result<Shipping> {
        val shipping = Shipping()
        return Result.success(shipping)
    }
    
    @GetMapping("/{id}/track")
    fun trackShipping(@PathVariable id: Long): Result<ShippingTrackVO> {
        val track = ShippingTrackVO()
        return Result.success(track)
    }
    
    @GetMapping("/tracking/{trackingNo}")
    fun trackByNumber(@PathVariable trackingNo: String): Result<ShippingTrackVO> {
        val track = ShippingTrackVO()
        return Result.success(track)
    }
    
    @PutMapping("/{id}/status")
    fun updateStatus(@PathVariable id: Long, @RequestParam status: Int): Result<Shipping> {
        val shipping = Shipping()
        return Result.success(shipping)
    }
    
    @PostMapping("/{id}/ship")
    fun shipOrder(
        @PathVariable id: Long,
        @RequestParam carrier: String,
        @RequestParam trackingNo: String
    ): Result<Shipping> {
        val shipping = Shipping()
        return Result.success(shipping)
    }
    
    @PostMapping("/{id}/deliver")
    fun markDelivered(@PathVariable id: Long): Result<Shipping> {
        val shipping = Shipping()
        return Result.success(shipping)
    }
    
    @GetMapping("/carriers")
    fun getCarriers(): Result<List<CarrierVO>> {
        val carriers = listOf<CarrierVO>()
        return Result.success(carriers)
    }
    
    @PostMapping("/calculate-fee")
    fun calculateFee(@RequestBody req: CalculateShippingFeeReq): Result<ShippingFeeVO> {
        val fee = ShippingFeeVO()
        return Result.success(fee)
    }
    
    @GetMapping("/{id}/history")
    fun getHistory(@PathVariable id: Long): Result<List<ShippingHistoryVO>> {
        val history = listOf<ShippingHistoryVO>()
        return Result.success(history)
    }
    
    @GetMapping("/pending")
    fun getPendingShipments(
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<Shipping>> {
        val result = PageResult<Shipping>()
        return Result.success(result)
    }
    
    @GetMapping("/in-transit")
    fun getInTransitShipments(
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<Shipping>> {
        val result = PageResult<Shipping>()
        return Result.success(result)
    }
    
    @GetMapping("/delivered")
    fun getDeliveredShipments(
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<Shipping>> {
        val result = PageResult<Shipping>()
        return Result.success(result)
    }
    
    @GetMapping("/statistics")
    fun getStatistics(): Result<ShippingStatisticsVO> {
        val statistics = ShippingStatisticsVO()
        return Result.success(statistics)
    }
    
    @PostMapping("/{id}/cancel")
    fun cancelShipping(@PathVariable id: Long): Result<Shipping> {
        val shipping = Shipping()
        return Result.success(shipping)
    }
    
    @GetMapping("/export")
    fun exportShipping(
        @RequestParam(required = false) startDate: String?,
        @RequestParam(required = false) endDate: String?
    ): Result<String> {
        val fileUrl = ""
        return Result.success(fileUrl)
    }
    
    @PostMapping("/batch-ship")
    fun batchShip(@RequestBody orderIds: List<Long>): Result<List<Shipping>> {
        val shippings = listOf<Shipping>()
        return Result.success(shippings)
    }
}

data class CreateShippingReq(
    val orderId: Long?,
    val carrier: String?,
    val trackingNo: String?,
    val receiverName: String?,
    val receiverPhone: String?,
    val address: String?
)

data class UpdateShippingReq(
    val carrier: String?,
    val trackingNo: String?,
    val receiverName: String?,
    val receiverPhone: String?,
    val address: String?
)

data class ShippingTrackVO(
    val trackingNo: String? = null,
    val carrier: String? = null,
    val status: Int? = null,
    val trackPoints: List<TrackPointVO>? = null
)

data class CarrierVO(
    val code: String? = null,
    val name: String? = null,
    val icon: String? = null,
    val website: String? = null
)

data class ShippingFeeVO(
    val fee: BigDecimal? = null,
    val carrier: String? = null,
    val estimatedDays: Int? = null
)

data class CalculateShippingFeeReq(
    val fromAddress: String?,
    val toAddress: String?,
    val weight: BigDecimal?,
    val carrier: String?
)

data class ShippingHistoryVO(
    val id: Long? = null,
    val action: String? = null,
    val description: String? = null,
    val location: String? = null,
    val operator: String? = null,
    val createdAt: String? = null
)

data class ShippingStatisticsVO(
    val totalShipments: Long? = null,
    val pendingShipments: Long? = null,
    val inTransitShipments: Long? = null,
    val deliveredShipments: Long? = null,
    val cancelledShipments: Long? = null
)
